soc/same54: Initial support

benpicco opened the pull request:
The Atmel SAME54 is a Cortex M4F with Ethernet that shares it's peripherals with the sam0 family. (This seems to be true for all new Atmel parts, it certainly is for the new SAML1x Cortex M23 parts).

The MCU is part of the SAMD5x/SAME5x family - the E parts are identical to the D ones, but come with an additional GMAC Ethernet peripheral.

This adds support for the MCU and the SAM E54 Xplained Pro development board.

I ported all existing sam0 peripherals use the new clock setup if available. 

What works:
- [x] **UART**
- [x] **SPI** - ~~depends on #14876~~
- [x] **I2C** - ~~depends on #14128~~
- [x] **GPIO**
- [x] **DMA**
- [x] **XOSC** 
- [x] **RTC** *when used as system timer, only non-tickless mode works*
- [x] **Watchdog**
- [x] **ADC** only tested with `adc_api` test
- [x] **USB**
- [x] **RNG**
- [ ] **Ethernet** *GMAC driver not implemented/ported yet*
- [x] **Timers** - ~~depends on #15105~~
- [x] **Flash**
- [ ] **QSPI** - *to be implemented*

requires https://github.com/zephyrproject-rtos/hal_atmel/pull/1

benpicco committed:
boards: atsame54_xpro: Add SAM E54 Xplained Pro Evaluation Kit

This adds support for the Atmel SAM E54 Xplained Pro Evaluation Kit.

Only basic functionality has been tested so far.

Signed-off-by: Benjamin Valentin <benpicco@googlemail.com>

benpicco committed:
watchdog: sam0: Add support for SAME54

The watchdog peripheral on SAME5x/SAMD5x MCUs is very simmilar
to the one found on the SAMD2x parts with only a few register
names changed.

Signed-off-by: Benjamin Valentin <benjamin.valentin@ml-pa.com>

benpicco committed:
flash: sam0: Add support for SAME54

This adds support for the flash peripheral found in the SAME5x/SAMD5x
line of MCUs.
The peripheral is very simmilar to the one found on SAMD2x with only
a few register names changed.

Signed-off-by: Benjamin Valentin <benjamin.valentin@ml-pa.com>

benpicco committed:
soc: atmel_sam0: Add SAME54

This adds supoprt for the Atmel SAME54 SoC.

The SAME5x/SAMD5x is a line of Cortex-M4F MCUs that share peripherals
with the sam0 Cortex-M0+ and saml1x Cortex-M23 parts.

Signed-off-by: Benjamin Valentin <benjamin.valentin@ml-pa.com>

benpicco committed:
gpio: sam0: Add support for PORTD

Some devices like SAME54 have an additional GPIO bank.

Signed-off-by: Benjamin Valentin <benjamin.valentin@ml-pa.com>

benpicco committed:
usb: sam0: Add support for SAME54

The USB peripheral is the same as in previous SAM0 MCUs, but the
clock setup is different.

Signed-off-by: Benjamin Valentin <benjamin.valentin@ml-pa.com>

benpicco committed:
spi: sam0: Add support for SAME54

The SPI SERCOM peripheral found on the SAMD5x/SAME5x is very much alike
the one found in previous SAM0 MCUs.

Only the clock setup is different.

Signed-off-by: Benjamin Valentin <benjamin.valentin@ml-pa.com>

benpicco committed:
timer: sam0_rtc_timer: Add support for SAME54

The RTC peripheral found in the SAMD5x/SAME5x MCUs is very
simmilar to the one found in existing sam0 devices with only
a few changes to register names and the clock source selection.

Signed-off-by: Benjamin Valentin <benpicco@googlemail.com>

benpicco committed:
pinmux: sam0: Add support for PORTD

Some devices like SAME54 have an additional GPIO port.

Signed-off-by: Benjamin Valentin <benjamin.valentin@ml-pa.com>

Sizurka reviewed (commented):
The parenthesis trick is very neat and I'm going to use it :)  But it might be better to put it with the other macro voodoo in the shared header so it can be reused:

[include/misc/util.h:387](https://github.com/zephyrproject-rtos/zephyr/blob/8e85d7d6c0f521ee8e19c2e58a793f32f43b8d89/include/misc/util.h#L387#L387)
```
#define UTIL_IS_PARENTHESIZED(x) UTIL_CHECK(UTIL_IS_PARENTHESIZED_ x, 0)
#define UTIL_IS_PARENTHESIZED_(...) 1, 1
```

You can also use some of the other macros in there to simply the usage later.  For example, I'm planning on changing the GPIO handling to something like:

```
#define SAM0_EIC_IRQ_CONNECT(n, _)					\
COND_CODE_1(UTIL_IS_PARENTHESIZED(DT_EIC_SAM0_IRQ(n)), ( {		\
	IRQ_CONNECT(DT_EIC_SAM0_IRQ(n), DT_EIC_SAM0_IRQ_PRIORITY(n),	\
		    sam0_eic_isr, DEVICE_GET(sam0_eic), 0);		\
	irq_enable(DT_EIC_SAM0_IRQ(n));					\
} ), () )

...

UTIL_LISTIFY(16, SAM0_EIC_IRQ_CONNECT, _);

```

So if you don't mind I'd like to add the DTS changing and however you decide to handle the macro definitions to the GPIO PR and I can just drop it if yours gets merged first.  Alternatively, I can just strip my current multi IRQ handling out and you can add something like the above to your changes for the EIC, whichever way you prefer.

benpicco commented:
Nice! I suppose it takes a super worm moon for this kind of macro magic to come forth.
Maybe it's easier to get this included as a separate pull request.

Sizurka commented:
Oh well, it was a nice dream.  I'll pull the parametric macros for the EIC out of `dts_fixup.h` as well, since that probably qualifies as "something weird".  That does mean you'll need 16 copy-paste definitions for the SAME54 one, though.

Really, it'd be nice if `scripts/dts/extract_dts_includes.py` (`extract/default.py` ?) generated kconfig style defines for the presence of all properties in the device, so there'd be a `#define DT_ATMEL_SAM0_UART_42001000_IRQ_0_PRESENT 1` or something.  That's probably too specialized to the SAM0 code generation needs, though.

benpicco commented:
Yes, I will indeed try to reduce the use of `dts_fixup.h` and switch to `DT_ATMEL_SAM0_UART_0_IRQ_0` et al. Since any re-ordering would be consistent when sticking to the defines from `generated_dts_board_unfixed.h`, I don't see how it would be a problem.

The parametric macros then can stay in the corresponding .c file

Sizurka commented:
Yeah I think it should be fine as long as you only reference devices by the label, it's just a bit odd when SERCOM3 ends up as DT_ATMEL_SAM0_UART_0 (since it was the first UART instance) while SERCOM1 is also DT_ATMEL_SAM0_SPI_0.  As long as it's consistent everything should be assigned correctly.  So I'll change the I2C code as well.

benpicco committed:
soc: atmel_sam0: Add SAMD51

This adds supoprt for the Atmel SAMD51 SoC.

The SAME5x/SAMD5x is a line of Cortex-M4F MCUs that share peripherals
with the sam0 Cortex-M0+ and saml1x Cortex-M23 parts.

Signed-off-by: Benjamin Valentin <benjamin.valentin@ml-pa.com>

Sizurka commented on dts/arm/atmel/samd5x.dtsi:
```diff
@@ -0,0 +1,212 @@
+		dma: dmac@4100A000 {
+			compatible = "atmel,sam0-dmac";
+			reg = <0x4100A000 0x50>;
+			interrupts = <31 0>, <32 0>, <33 0>, <34 0>;
```
I think you're missing IRQ 35 here (all DMA channels from 4..31).

benpicco replied:
good catch!

benpicco committed:
serial: sam0: Add support for SAME54

The SAME5x/SAMD5x MCUs share their SERCOM peripherals with the
samd2x and saml1x MCUs with only few registers changed.

Signed-off-by: Benjamin Valentin <benjamin.valentin@ml-pa.com>

benpicco committed:
soc: sam0: Add SERCOM fixup for samd5x

The SERCOMS on SAMD5x/SAME5x are connected to different MCLK
APBMASKs. There is no systematic way to tell whether a SERCOM
is connected to APBA, APBB, APBC or APBD, so rely on the
information from ASF instead of replicating it elsewhere.

This is needed for SPI, I2C and UART support on the SAMD5x/SAME5x
platform.

Signed-off-by: Benjamin Valentin <benjamin.valentin@ml-pa.com>

benpicco committed:
drivers: entropy: Add SAM0 entropy driver

This adds a driver for the True Random Number Generator found in some
Atmel SAM0 SoCs.
The Code is based on the driver for the SAM TRNG, but uses different
register and clock definitions.

Signed-off-by: Benjamin Valentin <benjamin.valentin@ml-pa.com>

dbkinder reviewed (approved):
+1 for docs

benpicco committed:
soc: atmel_sam0: Add SAME51

This adds supoprt for the Atmel SAME51 SoC.

The SAME5x/SAMD5x is a line of Cortex-M4F MCUs that share peripherals
with the sam0 Cortex-M0+ and saml1x Cortex-M23 parts.

Signed-off-by: Benjamin Valentin <benpicco@googlemail.com>

benpicco committed:
soc: atmel_sam0: Add SAME53

This adds supoprt for the Atmel SAME53 SoC.

The SAME5x/SAMD5x is a line of Cortex-M4F MCUs that share peripherals
with the sam0 Cortex-M0+ and saml1x Cortex-M23 parts.

Signed-off-by: Benjamin Valentin <benpicco@googlemail.com>

benpicco committed:
interrupt_controller: sam0: Add support for SAME54

The EIC/PORT peripheral works very much alike the one in
previous sam0 parts.

Signed-off-by: Benjamin Valentin <benpicco@googlemail.com>

benpicco committed:
i2c: sam0: Add support for SAME54

It shares the same peripheral with other sam0 parts while differing
in the clock config.

Signed-off-by: Benjamin Valentin <benpicco@googlemail.com>

Sizurka commented on dts/arm/atmel/samd5x.dtsi:
```diff
@@ -0,0 +1,317 @@
+			interrupts = <118 0>, <119 0>;
+			label = "ADC_0";
+
+			/* FIXME: is this stil true?
```
This looks like a bit of an odd case: the SAMD5x manual says fGCLK_ADC max is 100 MHz (table 54-8, in section 54.6, p 2020), while the F_ADC is 16 MHz (table 54-24, section 54.10.4, p 2031).  So though you can run it faster (i.e. 16 MHz after division for the rated 1 MSPS), clocking it off the full 120 MHz GCLK0 is technically out of spec, as I read it.

benpicco replied:
Thank you! I hadn't really looked into this yet, just wanted something to compile your ADC driver with 😄 

Since currently I only configure 2 clocks, the main clock (120MHz, but can be set in KConfig) and a 32kHz clock, I suppose I should default to the latter here if I don't want to introduce a separate clock just for ADC.

I suppose this will be much cleaner with #15585

benpicco committed:
drivers: dma: sam0: fix DMA to peripheral transfer on SAMD5x

We need to select TRIGACT_BLOCK for DMA to peripheral to work.

Signed-off-by: Benjamin Valentin <benjamin.valentin@ml-pa.com>

Sizurka commented on drivers/flash/flash_sam0.c:
```diff
@@ -146,8 +165,12 @@ static int flash_sam0_write_page(struct device *dev, off_t offset,
-
+#else
+	/* TODO: is erase page the right thing to do here? */
+	NVMCTRL->CTRLB.reg = NVMCTRL_CTRLB_CMD_EP | NVMCTRL_CTRLB_CMDEX_KEY;
```
I think you're looking for erase block (CMD_EB), "blocks" in Microchip speak here are the the minimum erasure unit for most of the flash ("page" in Microchip speak is the program unit, which looks like it's only available on the AUX flash (table 25-2)).

Additionally, from the line below that table, I think you need to adjust "PAGES_PER_ROW" to 16.  That is, I think SAMD5x just replaced "row" with "block" terminology wise.

benpicco replied:
Again, thank you!

Sizurka commented on dts/arm/atmel/samd5x.dtsi:
```diff
@@ -0,0 +1,281 @@
+	flash0: flash@0 {
+		compatible = "soc-nv-flash";
+		label = "FLASH_0";
+		write-block-size = <64>;
```
I am less sure about this, but I think this should be 32:  `FLASH_PAGE_SIZE` is 512 for SAMD5x in the ASF headers, and with 16 pages/block = 32 bytes/page (I really wish they'd just spell this out somewhere, but I didn't find it any any obvious locations).

Sizurka replied:
Actually on further inspection: this is `FLASH_PAGE_SIZE` is 64 on SAMD2x (i.e. at least the terminology is consistent, I guess).  So this should probably be 512 (which makes the erasure unit 8KiB, which is consistent with the ASF headers).

So I think the right way to do this, is to set `ROW_SIZE` to `NVMCTRL_ROW_SIZE` or `NVMCTRL_BLOCK_SIZE` (depending on which is defined from ASF), then calculate `PAGES_PER_ROW` from `FLASH_PAGE_SIZE / ROW_SIZE`.

Unfortunately, I don't see any definition that could be used for `LOCK_REGIONS`, which looks like it's 32 on SAMD5x.

benpicco replied:
Thank you! With this `samples/drivers/flash_shell` is mostly working.
(I can read/write and erase pages once, when I try to erase a second page, it crashes at `NVMCTRL->CTRLB.reg = NVMCTRL_CTRLB_CMD_UR | NVMCTRL_CTRLB_CMDEX_KEY;` - I'll still have to investigate this.

I've skimmed a few more data sheets too and it looks like `LOCK_REGIONS` is something for the device tree:

MCU | LOCK_REGIONS
------|------------------
SAMD2x | 16
SAML2x  | 16
SAML10  |  2
SAML11  |  4
SAMD5x  | 32

Sizurka commented on drivers/gpio/gpio_sam0.c:
```diff
@@ -291,3 +291,15 @@ DEVICE_AND_API_INIT(gpio_sam0_2, DT_ATMEL_SAM0_GPIO_PORT_C_LABEL,
+
+static const struct gpio_sam0_config gpio_sam0_config_3 = {
+	.regs = (PortGroup *)DT_ATMEL_SAM0_GPIO_PORT_D_BASE_ADDRESS,
+};
```
needs `.id = 3` for interrupt support.

benpicco replied:
Good catch!

Sizurka commented on drivers/usb/device/usb_dc_sam0.c:
```diff
@@ -181,6 +204,7 @@ int usb_dc_attach(void)
+#endif /* !MCLK */
 
 	/* Configure */
 	regs->CTRLA.bit.SWRST = 1;
```
Don't the interrupts 2 and 3 also need to be connected (TRCPT0 and TRCPT1).  It looked like you have them in the DTS already.

benpicco replied:
Good catch!
The `samples/subsys/usb/cdc_acm` sample is still not working, I'll have to investigate this further.

Sizurka commented on drivers/dma/dma_sam0.c:
```diff
@@ -149,7 +149,7 @@ static int dma_sam0_config(struct device *dev, u32_t channel,
 	} else {
 		/* One peripheral trigger per beat */
-		chcfg->CHCTRLA.reg = DMAC_CHCTRLA_TRIGACT_BURST |
+		chcfg->CHCTRLA.reg = DMAC_CHCTRLA_TRIGACT_BLOCK |
```
The burst length setting code below should also be removed, since it won't do anything if not using burst mode (which apparently doesn't work, oh well).  That is, all the code below related to burst length can also be removed, since the defaults (0) are fine in block mode.

benpicco replied:
OK, removed.

This makes the code even more similar between the two versions of the peripheral.

benpicco committed:
flash: sam0: Define LOCK_REGIONS in dts

The amount of lock regions differs between different sam0 MCUs.

	saml10:  2
	saml11:  4
	samd2x:	16
	saml2x: 16
	samd5x: 32

ASF does not provide a definition for this, so create a new one
in dts.

Signed-off-by: Benjamin Valentin <benjamin.valentin@ml-pa.com>

Sizurka commented on drivers/flash/flash_sam0.c:
```diff
@@ -339,9 +343,8 @@ static int flash_sam0_write_protection(struct device *dev, bool enable)
-	     offset += LOCK_REGION_SIZE) {
-		*FLASH_MEM(offset) = 0U;
+	for (offset = 0; offset < CONFIG_FLASH_SIZE * 1024;  offset += LOCK_REGION_SIZE) {
+		NVMCTRL->ADDR.reg = offset;
```
I believe this should be `NVMCTRL->ADDR.reg = offset + CONFIG_FLASH_BASE_ADDRESS;`.

As I interpret it, the whole `*FLASH_MEM(offset) = 0U;` thing that's done is an (ab)use of the note about the address register "This register is also automatically updated when writing to the page buffer.", so it's trying to do a write at the address which resolves to `offset + CONFIG_FLASH_BASE_ADDRESS`, but not actually filling/issuing the command so the only effect "should" be the address update.  It's not clear to me why this idiom was used instead of the more direct just loading the address then sending the command (like you changed this one to).

benpicco replied:
You are of course right.
The reason this still worked was that `CONFIG_FLASH_BASE_ADDRESS` is 0 - fixed.

Sizurka commented on soc/arm/atmel_sam0/common/soc_samd5x.c:
```diff
@@ -0,0 +1,108 @@
+	/* enable the Cortex M Cache Controller */
+	CMCC->CTRL.bit.CEN = 1;
+
+	z_clearfaults();
```
The `z_clearfaults();` can be dropped since #16169 has been merged.

benpicco committed:
gpio: sam0: provide PORT_PMUX_PMUXE_A_Val define

SAMD5x/SAME5x header files do not provide this define anymore.
On SAMD2x it was 0, this is still valid.

Signed-off-by: Benjamin Valentin <benpicco@googlemail.com>

benpicco commented:
Rebased and re-run the tests.
As the vendor files have been moved to a separate repository, https://github.com/zephyrproject-rtos/hal_atmel/pull/1 is needed for this now.

galak commented:
Shouldn't this update west.yaml to point to https://github.com/zephyrproject-rtos/hal_atmel/pull/1

benpicco commented:
@galak I didn't know I can do this! Thank you.

benpicco committed:
west.yml: update hal_atmel for SAMD5x/SAME5x

Update hal_atmel to include vendor files for SAMD5x/SAME5x.

Signed-off-by: Benjamin Valentin <benjamin.valentin@ml-pa.com>

galak commented on west.yml:
```diff
@@ -21,12 +21,15 @@ manifest:
     - name: hal_atmel
-      revision: 04ff67a0826a51041e51034faf8fc4d3eeacd846
+      remote: benpicco
+      revision: 32ce0ddd1a3e0512de9822693e334d7090776ec0
```
This should just be:

`revision: pull/1/head`

benpicco replied:
Alright!
I didn't intent to commit this to master though 😉

galak commented:
> Can we reduce this PR to have base SoC & board support, and than have the driver updates as separate PRs?

I suggest this so that the driver PRs can hopefully get reviewed by proper driver maintainers.

benpicco commented:
Alright then, I'll split this up.

benpicco commented:
@galak I've split it into individual PRs as requested.

stephanosio commented:
> Ethernet GMAC driver not implemented/ported yet
@benpicco Do you plan to/are you already working on this?

benpicco commented:
@stephanosio no i don't have any concrete plans for that.

stephanosio commented:
> @stephanosio no i don't have any concrete plans for that.

Ok, I will look into it then. At first glance, it seems to be using the same GMAC core as SAM7.
